### PR TITLE
kie-issues#117: BPMN Editor: Enable auto population of DMN fields in Business Rule Task

### LINKED ISSUE #117: BPMN Editor: Enable auto population of DMN fields in Business Rule Task
When selecting DMN as rule language, the following fields should be fetched from a selected file name (New Field) and the following fields should be populated

Namespace, Decision Name (Dropdown) and DMN Model Name

The new field File Name should be marshalled and unmarshalled accordingly in BPMN Model

image

### PR DESCRIPTION
**Pull request by inodeman**
Hi @tiagobento @porcelli here is the PR for the new functionality. 
@treblereel @yesamer can you review this PR?
I have uploaded builds to: https://ibm.box.com/s/rqvab9c51hsqrzjvip031fyalbrn5t43
For the online_editor just unzip the file and run pnpm start and head to https://localhost:9001
Note: It will find the dmn files saved to the same directory as the bpmn file being tested.
For the Vs code extension I have included some sample files, File5.dmn is empty just to test it will not creash the logic.
I have tested several edge cases and fixed some issues that came up, let me know if you find any

Closes https://github.com/kiegroup/kie-issues/issues/117

### CODE REVIEW

**Review by yesamer — COMMENTED** [review]
Very nice feature, @inodeman !

I manually tested it, and I noticed these two cases:

**Traffic Violation** 
An "Amount" decision is detected but is not present in the DMN DRG
Screenshot 2023-03-15 at 10 28 16
Screenshot 2023-03-15 at 10 28 01

**Loan Pre-Qualification** 
The "Loan Pre-Qualification" decision is missing 
Screenshot 2023-03-15 at 10 49 01
Screenshot 2023-03-15 at 10 52 00

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DMNParsingUtils.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        }
+        return dmnDecisionNames;
+
+    }
+
+    public static String parseDMNModelName(final String content) {
+        return parseField("name=\"", "\"", content);
+    }
+
+}
```
Comment (tiagobento) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DMNParsingUtils.java]: @inodeman Ok, so I think this logic needs to be improve, and needs to live at the `dmn-language-service` that @ljmotta started a while ago. We should be using a true XML parsing.

The `dmn-language-service` library is still on a very early stage, so let's say it's a "best effort" currently. That's okay, you can use the same approach we used for the other methods it already has.

To integrate `dmn-language-service` with the BPMN Editor, you can use the same approach that we did for the PMML Marshaller on the DMN Editor. The PMML Marshaller is a TypeScript library that gets injected inside the DMN Editor. See `packages/kie-bc-editors/src/dmn/envelope/DmnEditorFactory.ts`

We should do exactly the same for the DMN Language Service on the BPMN Editor. On the GWT-side, we use JsInterop to consume this library and get the necessary information from the DMN files. I don't think the `dmn-language-service` library will have all the capabilities you need for this PR, so creating it should be part of the PR.

Comment (tiagobento) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DMNParsingUtils.java]: A little more context:

`dmn-language-service` was created so that DMN Runner only deals with DMN files that are associated with the DMN it's running. The methods there are more related to the included models part. Here, we're parsing a single DMN file and its attributes, so new methods will have to be created.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java [hunk 2]
```diff
@@ -141,11 +147,14 @@ public void show(final Command callback) {
             }
         }
     }
-
     @SuppressWarnings("unchecked")
     public boolean executeUpdateProperty(final Element<? extends Definition<?>> element,
                                          final String fieldName,
                                          final Object value) {
+
+        if (fieldName.equals("executionSet.fileName")) { // Special Condition to enable DMN Auto population on BusinessRuleTask
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java]: @inodeman I'd try to avoid this kind of condition here to keep it as much generic as possible. Also I'm worried that the `executionSet.fileName` expression could be used in other stunner models in the future and this could carry issues.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/BusinessRuleTaskExecutionSet.java [hunk 3]
```diff
@@ -68,14 +69,30 @@ public class BusinessRuleTaskExecutionSet implements BPMNPropertySet {
...
+            afterElement = "fileName"
     )
     @Valid
     private Namespace namespace;
 
     @Property
     @FormField(
-            afterElement = "namespace"
+            afterElement = "namespace",
+            type = ComboBoxFieldType.class
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/BusinessRuleTaskExecutionSet.java:91]: Theres a mechanism to make a dropdown depend on another one by config by adding the `relatedField` param to the config, you can look at: https://github.com/kiegroup/kie-tools/blob/main/packages/stunner-editors/kie-wb-common-forms/kie-wb-common-forms-core/kie-wb-common-forms-fields/src/main/java/org/kie/workbench/common/forms/fields/shared/fieldTypes/basic/selectors/listBox/definition/StringListBoxFieldDefinition.java#L52

IDK if that could help, but just dropping it here.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java [hunk 4]
```diff
@@ -0,0 +1,224 @@
...
+import org.kie.workbench.common.stunner.core.client.event.util.FileNameElementSetterEvent;
+import org.kie.workbench.common.stunner.core.graph.Element;
+import org.kie.workbench.common.stunner.core.graph.content.definition.Definition;
+import org.kie.workbench.common.stunner.forms.client.event.FormFieldChanged;
+import org.kie.workbench.common.stunner.forms.client.event.FormPropertiesOpened;
+import org.kie.workbench.common.stunner.forms.client.event.RefreshFormPropertiesEvent;
+import org.uberfire.client.promise.Promises;
+
+@Dependent
+public class FileNameFormProvider implements SelectorDataProvider {
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java:28]: I think this class is mixing three behaviours that should be spread in three different components:
1.- the`SelectorDataProvider ` that is actually providing the dropdown values.
2.- the class that interacts with the `ResourceContentService ` and is able to read the dmn and it's fields.
3.- the class that has knowledge about which is the current edited element and is able to modify the `namespace`,  `decission_name` and `dmn_model_name`

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+                + BusinessRuleTaskExecutionSet.DMN_MODEL_NAME, element, dmnModelName);
+    }
+
+    public List<String> getDecisionName() {
+        return decisionName;
+    }
+
+    public void onFormFieldChanged(@Observes FormFieldChanged formFieldChanged) {
+
+        String fileName = formFieldChanged.getValue().toString();
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java]: I'd preffer this kind of logic in a specific component that knows which is the current edited element (and if it's a BussinessTask), so we keep the `SelectorDataProvider` decoupled from any other.

This same class could be used for the `DecissionNameDataProvider` to get the businesstask element

**Review by pefernan — CHANGES_REQUESTED** [review]

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java (hunk 2 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java]: Hi @pefernan refactored it according to our call

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/BusinessRuleTaskExecutionSet.java (hunk 3 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-api/src/main/java/org/kie/workbench/common/stunner/bpmn/definition/property/task/BusinessRuleTaskExecutionSet.java:91]: Hi @pefernan I removed the reference to FileNameFormProvider and used a singleton instead, hope this solves it

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java (hunk 4 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java:28]: Hi @pefernan refactored it as requested

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java (hunk 5 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/dataproviders/FileNameFormProvider.java]: Hi @pefernan refactor done

**Comment by inodeman** [comment]
Hi @yesamer changed the parsing strategy as commented by @tiagobento and now the issues you mention do not happen. Tiago, added the code to dmn-language-service as requested, as well as refactoring mentioned by @pefernan PR is ready for re-review

File: packages/kie-bc-editors/src/bpmn/envelope/exposedInteropApi/DMNEditorMarshallerExposedInteropApi.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * This API is exposed from the Envelope to be consumed on Java code.
+ */
+export interface DMNEditorMarshallerExposedInteropApi {
+  /**
+   * It performs the marshall of a xmlContent retrieved from a DMN file returning a DMNDocumentData
+   * * @param xmlContent
+   */
+  getDMNDocumentData(dmnContent: string): DMNDocumentData;
+}
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/bpmn/envelope/exposedInteropApi/DMNEditorMarshallerExposedInteropApi.ts]: Let's not use all-caps acronyms and do `DmnEditorMarshallerExposedInteropApi` and `getDmnDocumentData`.

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts [hunk 7]
```diff
@@ -26,6 +36,23 @@ export class BpmnEditorFactory implements EditorFactory<BpmnEditor, BpmnEditorCh
     ctx: KogitoEditorEnvelopeContextType<BpmnEditorChannelApi>,
     initArgs: EditorInitArgs
   ): Promise<BpmnEditor> {
+    const exposedInteropApi: CustomWindow["envelope"] = {
+      dmnEditorMarshallerService: new DmnLanguageService({
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: This name could be simply `dmnLanguageService`.

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts [hunk 8]
```diff
@@ -26,6 +36,23 @@ export class BpmnEditorFactory implements EditorFactory<BpmnEditor, BpmnEditorCh
...
+        getDmnImportedModel: () =>
+          new Promise((res) =>
+            res({
+              relativePath: "",
+              content: "",
+            })
+          ),
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: No need to create a new Promise using its constructor, you can do `Promise.resolve({ relativePath: "", content: "" })`.

However, in this case, this is a "mock" implementation, so we should eiher:
1. Add a comment saying why we're not properly implementing this; or
2. Implement it properly :D

File: packages/dmn-language-service/src/DMNDecision.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+ * limitations under the License.
+ */
+
+export class DMNDecision {
+  public readonly name: string;
+
+  constructor(name: string) {
+    this.name = name;
+  }
+}
```
Comment (tiagobento) [inline packages/dmn-language-service/src/DMNDecision.ts]: `DmnDecision`

File: packages/dmn-language-service/src/DMNDocumentData.ts [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { DMNDecision } from "./DMNDecision";
+
+export class DMNDocumentData {
```
Comment (tiagobento) [inline packages/dmn-language-service/src/DMNDocumentData.ts]: `DmnDocumentData`

File: packages/dmn-language-service/src/index.ts [hunk 11]
```diff
@@ -67,6 +79,36 @@ export class DmnLanguageService {
...
+      const dmnDecision = new DMNDecision(decisionName);
+      if (dmnDecision) {
+        dmnDecisionsArray.push(dmnDecision);
+      }
+    });
+    return new DMNDocumentData(
+      namespace !== null ? namespace : "",
+      dmnModelName !== null ? dmnModelName : "",
+      dmnDecisionsArray
+    );
```
Comment (tiagobento) [inline packages/dmn-language-service/src/index.ts]:
```suggestion
const decisions = Array.from(dmnDecisionsContent)
      .map((decision) => decision.getAttribute(DECISION_NAME_ATTRIBUTE))
      .flatMap(decisionName => decisionName ? [new DmnDecision(decisionName)] : []);

    return new DmnDocumentData(
      namespace ?? "",
      dmnModelName ?? "",
      decisions,
    );
```

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts [hunk 12]
```diff
@@ -26,6 +36,23 @@ export class BpmnEditorFactory implements EditorFactory<BpmnEditor, BpmnEditorCh
...
+      }),
+    };
```
Comment (tiagobento) [inline packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: In general, although this is 100% correct, I'm not sure I like the idea of coupling the DMN LS with its usage on the BPMN Editor. If you don't create an indirection here, it means that the BPMN Editor is 100% directly coupled with the types that the DMN LS API.

I think the DMN LS must be more independent, so I would create intermediate types here (can be 1:1 copy from what is currently there). Please see the `exposedInteropApi` directory on `kie-bc-editors` for examples. 

Note that some features exposed to GWT are NOT following this practice, like PMML Marshaller on DMN Editor and `ResourceContentOptions` on `ResourceContentExposedInteropApi`. That's probably my fault, though. Don't be like me! :D

**Review by tiagobento — CHANGES_REQUESTED** [review]
@inodeman Excellent work! Glad to see you touching the TS world too! :)

Left some comments inline, please take a look at them! 

I didn't even look at the Java part, as @pefernan will do a much better job than I at it :)

File: packages/kie-bc-editors/src/bpmn/envelope/exposedInteropApi/DMNEditorMarshallerExposedInteropApi.ts (hunk 6 above)
Comment (inodeman) [reply packages/kie-bc-editors/src/bpmn/envelope/exposedInteropApi/DMNEditorMarshallerExposedInteropApi.ts]: Hi @tiagobento changed

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts (hunk 7 above)
Comment (inodeman) [reply packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: Hi @tiagobento changed

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts (hunk 8 above)
Comment (inodeman) [reply packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: Hi @tiagobento added the comment as discussed

File: packages/dmn-language-service/src/DMNDecision.ts (hunk 9 above)
Comment (inodeman) [reply packages/dmn-language-service/src/DMNDecision.ts]: Hi @tiagobento changed

File: packages/dmn-language-service/src/DMNDocumentData.ts (hunk 10 above)
Comment (inodeman) [reply packages/dmn-language-service/src/DMNDocumentData.ts]: Hi @tiagobento changed

File: packages/dmn-language-service/src/index.ts (hunk 11 above)
Comment (inodeman) [reply packages/dmn-language-service/src/index.ts]: Hi @tiagobento changed

File: packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts (hunk 12 above)
Comment (inodeman) [reply packages/kie-bc-editors/src/bpmn/envelope/BpmnEditorFactory.ts]: Hi @tiagobento changed

**Comment by inodeman** [comment]
Hi @tiagobento @pefernan I made changes according to your latest comments, let me know if ok. I found an issue that I am trying to fix, when you select a decision and you select a new file name, sometimes the value stays along with the new values from the file, it does not happen all the time, but will fix it next week, but wanted to get your ok on the code. Thanks

**Review by tiagobento — APPROVED** [review]
All looks good for me! Thanks @inodeman!

@pefernan I'm deferring the Java part to you :)

**Comment by tiagobento** [comment]
@inodeman Please keep in mind that we can only merge this if the CI is green.

**Review by yesamer — COMMENTED** [review]
Thank you @inodeman, I confirm you the missing decision nodes are now present.

Can you please address these minor changes?

1. Can you please alphabetically order the results?
Screenshot 2023-03-28 at 15 25 34

2. Can you please reset the Decision Name selection box when changing the DMN file? This is in order to avoid the following behavior:

https://user-images.githubusercontent.com/16005046/228255765-f4774509-49d9-464e-aada-a49a2a3babdf.mov

Thank you!

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java [hunk 13]
```diff
@@ -141,7 +141,6 @@ public void show(final Command callback) {
             }
         }
     }
-
```
Comment (yesamer) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java]: @inodeman Typo?

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        resourceContentService
+                .list(fileMatcher)
+                .then(paths -> {
+                    if (paths.length != 0) {
+                        fileNames.clear();
+                        for (String path : paths) {
+                            fileNames.put(path, path);
+                        }
+                        // Update forms in case the fileName field has a value, it needs to add it and the results returned
+                        refreshForms();
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java:73]: Do we need to refresh the forms here?

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+                                     final DmnLanguageServiceServiceProducer dmnLanguageServiceServiceProducer) {
+        this.resourceContentService = resourceContentService;
+        this.promises = promises;
+        this.refreshFormPropertiesEvent = refreshFormPropertiesEvent;
+        this.sessionManager = sessionManager;
+        this.dmnLanguageServiceServiceProducer = dmnLanguageServiceServiceProducer;
+    }
+
+    public void init(final BusinessRuleTaskChangeHandler businessRuleTaskChangeHandler) {
+        this.businessRuleTaskChangeHandler = businessRuleTaskChangeHandler;
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java]: I'd prefer not to have a refernce to `BusinessRuleTaskChangeHandler` here since `BusinessRuleTaskChangeHandler` already has a reference of `DmnResourceContentFetcher ` .

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java [hunk 16]
```diff
@@ -0,0 +1,112 @@
...
+    }
+
+    public void fetchFile(String fileName) {
+
+        resourceContentService
+                .get(fileName)
+                .then(content -> {
+                    DmnLanguageServiceApi produce = dmnLanguageServiceServiceProducer.produce();
+                    DmnDocumentData dmnDocumentData = produce.getDmnDocumentData(content);
+                    businessRuleTaskChangeHandler.setData(dmnDocumentData);
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java]: I'd prefer not to do it like this, I think the this `fetchFile` method could receive a `Consumer<DmnDocumentData>` to be called as a callback inside of the then, So `BusinessRuleTaskChangeHandler` can pass it with to update the values?

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+                    refreshForms();
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java:91]: Since we are going to update the form model inside `BusinessRuleTaskChangeHandler` I think this isn't needed.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+        this.businessRuleTask = businessRuleTask;
+    }
+
+    @Override
+    public void onFieldChange(String fieldName, Object newValue) {
+        if (fieldName.equals(FILE_NAME_FIELD)) {
+            String fileName = (String) newValue;
+
+            if (dmnResourceContentFetcher.getFileNames().get(newValue) != null) {
+                dmnResourceContentFetcher.fetchFile(fileName);
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java]: I'd pass a `Consumer<DmnDocumentData >` as a second parameter here to update the value.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+    protected final static String FILE_NAME_FIELD = BusinessRuleTask.EXECUTION_SET + "." + BusinessRuleTaskExecutionSet.FILE_NAME;
+
+    private BusinessRuleTask businessRuleTask;
+
+    private final DmnResourceContentFetcher dmnResourceContentFetcher;
+
+    @Inject
+    public BusinessRuleTaskChangeHandler(final DmnResourceContentFetcher dmnResourceContentFetcher) {
+        this.dmnResourceContentFetcher = dmnResourceContentFetcher;
+        dmnResourceContentFetcher.init(this); // Enable to change this
```
Comment (pefernan) [inline packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java]: I think this isn't needed.

**Review by pefernan — CHANGES_REQUESTED** [review]
@inodeman thanks! just added a few comments. My only concern is the way `BusinessRuleTaskChangeHandler` updates the model values when `DmnResourceContentFetcher` fetches the Dmn content, I think it would be better if we use a callback mechanism in the `fetchFile` method instead of keeping a reference of `BussinesRuleTaskChangeHandler`, this way `DmnResourceContentFetcher` is decoupled of the change handler and could be used in a more generic way.

Also think `DmnResourceContentFetcher` doesn't need to force updating the forms since `BussinesRuleTaskChangeHandler` has access to model bound to the form.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java (hunk 13 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-forms/kie-wb-common-stunner-forms-client/src/main/java/org/kie/workbench/common/stunner/forms/client/widgets/FormsCanvasSessionHandler.java]: Hi @yesamer yes, sorry, changed it back

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java (hunk 14 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java:73]: Hi @pefernan yes, I made this change for the following scenario. 1) The user has previously selected a filename in a saved bpmn file, the first time it loads it shows the file name but the other possible values are not present, that is why I am refreshing the forms here.

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java (hunk 15 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java]: Hi @pefernan will change it for callback mechanism

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java (hunk 16 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java]: Hi @pefernan will change it

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java (hunk 17 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java:91]: Hi @pefernan will change it

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java (hunk 18 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java]: Hi @pefernan will change it

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java (hunk 19 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/forms/changeHandlers/businessRuleTask/BusinessRuleTaskChangeHandler.java]: Hi @pefernan will remove the init call but the injection of the resource fetcher will be needed

File: packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java (hunk 17 above)
Comment (inodeman) [reply packages/stunner-editors/kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/util/DmnResourceContentFetcher.java:91]: Hi @pefernan as discusses, seems it is needed, if I don´t call it, it does not update until I close properties tab and reopen it

**Comment by inodeman** [comment]
Hi @pefernan ready for final review, made the last changes, also now it sorts as @yesamer requested.

**Comment by inodeman** [comment]
Hi @ederign, @paulovmr can you review?

**Comment by paulovmr** [comment]
Hi @inodeman , will leave this one with @pefernan and @tiagobento :)

**Review by pefernan — APPROVED** [review]
@inodeman ok I understand why you need to fire the event. Looks like there's an issue on the select box renderer that doesn't update the dropdown values when updating the proxy.

+1 to merge!

**Review by yesamer — APPROVED** [review]
@inodeman  Thank you!
